File: clojure/bob/src/bob.clj
(ns bob)

(defn shouting?
  "Message is considered shouting if it has at least one letter and is all upper case."
  [message] (and (re-find #"[A-Z]" message)
                 (= message (clojure.string/upper-case message))))

(defn question? [message] (clojure.string/ends-with? message "?"))

(defn silence? [message] (= " " (apply str (into #{} message))))

(defn numeric? [message] (re-find #"\d+" message))

(defn response-for [message]
  (cond
    (empty? message) "Fine. Be that way!"
    (silence? message) "Fine. Be that way!"
    (shouting? message) "Whoa, chill out!"
    (question? message) "Sure."
    :else "Whatever."))

File: python/bob/bob.py
import re

def numeric(string):
    return "".join(re.findall("[\d?]", string)) == string

def any(coll):
    return len(coll) > 0

def shouting(string):
    return any(re.findall("[a-zA-Z]", string)) and string.upper() == string

def asking(string):
    return string[-1] == "?"

def empty(string):
    return len(string) == 0

def hey(what):
    what = what.rstrip()
    if empty(what):
        return "Fine. Be that way!"
    elif shouting(what):
        return "Whoa, chill out!"
    elif asking(what):
        return "Sure."
    else:
        return "Whatever."
